### PR TITLE
fix ARM64 test panics

### LINKED ISSUE #346: arm64@go1.16+: tests panic
#### System information

```
[kde@plasma-mobile core-geth]$ git rev-parse HEAD
4d99a68451142ca67d9408a7b8f98e52cf439fc8

[kde@plasma-mobile core-geth]$ go version
go version go1.16.2 linux/arm64

[kde@plasma-mobile core-geth]$ uname -a
Linux plasma-mobile 5.11.3-1-MANJARO-ARM #1 SMP Thu Mar 4 17:54:13 UTC 2021 aarch64 GNU/Linux

[kde@plasma-mobile core-geth]$ df -h
Filesystem      Size  Used Avail Use% Mounted on
dev             1.4G     0  1.4G   0% /dev
run             1.5G  1.4M  1.5G   1% /run
/dev/mmcblk2p2   29G  9.6G   18G  36% /
tmpfs           1.5G     0  1.5G   0% /dev/shm
tmpfs           4.0M     0  4.0M   0% /sys/fs/cgroup
tmpfs           1.5G  1.5G   28M  99% /tmp
/dev/mmcblk2p1  214M   37M  177M  18% /boot
tmpfs           300M   48K  300M   1% /run/user/1000

[kde@plasma-mobile ~]$ cat /proc/cpuinfo
processor       : 0
BogoMIPS        : 48.00
Features        : fp asimd evtstrm aes pmull sha1 sha2 crc32 cpuid
CPU implementer : 0x41
CPU architecture: 8
CPU variant     : 0x0
CPU part        : 0xd03
CPU revision    : 4

processor       : 1
BogoMIPS        : 48.00
Features        : fp asimd evtstrm aes pmull sha1 sha2 crc32 cpuid
CPU implementer : 0x41
CPU architecture: 8
CPU variant     : 0x0
CPU part        : 0xd03
CPU revision    : 4

processor       : 2
BogoMIPS        : 48.00
Features        : fp asimd evtstrm aes pmull sha1 sha2 crc32 cpuid
CPU implementer : 0x41
CPU architecture: 8
CPU variant     : 0x0
CPU part        : 0xd03
CPU revision    : 4

processor       : 3
BogoMIPS        : 48.00
Features        : fp asimd evtstrm aes pmull sha1 sha2 crc32 cpuid
CPU implementer : 0x41
CPU architecture: 8
CPU variant     : 0x0
CPU part        : 0xd03
CPU revision    : 4
```

#### Expected behaviour

Tests pass.

#### Actual behaviour

Tests at `cmd/geth` panic.

#### Steps to reproduce the behaviour

This branch shows that Go versions 1.16+ are correlated with the panic, while <=1.15.10 appear unaffected.
https://github.com/meowsbits/core-geth/tree/arm64-test-panic
> :x: https://travis-ci.org/github/meowsbits/core-geth/builds/763558487

#### Trace

```
[kde@plasma-mobile core-geth]$ make test
[...]
?       github.com/ethereum/go-ethereum/cmd/evm [no test files]
?       github.com/ethereum/go-ethereum/cmd/evm/internal/compiler       [no test files]
?       github.com/ethereum/go-ethereum/cmd/evm/internal/t8ntool        [no test files]
ok      github.com/ethereum/go-ethereum/cmd/faucet      0.261s
unexpected fault address 0x4ce24b1b59d629e6
fatal error: fault
[signal SIGSEGV: segmentation violation code=0x1 addr=0x4ce24b1b59d629e6 pc=0x67fff0]

goroutine 207 [running]:
runtime.throw(0x132cf7f, 0x5)
        /usr/local/go/src/runtime/panic.go:1117 +0x54 fp=0x40005f3030 sp=0x40005f3000 pc=0x44c534
runtime.sigpanic()
        /usr/local/go/src/runtime/signal_unix.go:741 +0x230 fp=0x40005f3070 sp=0x40005f3030 pc=0x4640e0
github.com/golang/snappy.encodeBlock(0x40004aa602, 0x12f6, 0x12f6, 0x40002cf500, 0x1023, 0x13cd, 0x40000b4001)
        /home/kde/go/pkg/mod/github.com/golang/snappy@v0.0.2-0.20200707131729-196ae77b8a26/encode_arm64.s:666 +0x360 fp=0x40005fb110 sp=0x40005f3080 pc=0x67fff0
github.com/golang/snappy.Encode(0x40004aa600, 0x12f8, 0x12f8, 0x0, 0x0, 0x0, 0x2, 0x4, 0x40002b31e8)
        /home/kde/go/pkg/mod/github.com/golang/snappy@v0.0.2-0.20200707131729-196ae77b8a26/encode.go:39 +0x17c fp=0x40005fb160 sp=0x40005fb110 pc=0x67ec7c
github.com/syndtr/goleveldb/leveldb/table.(*Writer).writeBlock(0x40000c0240, 0x40000c0298, 0x2, 0x400051a060, 0x40007e5441, 0x28, 0x31abbf)
        /home/kde/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/table/writer.go:171 +0xb0 fp=0x40005fb1e0 sp=0x40005fb160 pc=0x6fa0f0
github.com/syndtr/goleveldb/leveldb/table.(*Writer).finishBlock(0x40000c0240, 0x40007e5441, 0x28)
        /home/kde/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/table/writer.go:222 +0x4c fp=0x40005fb230 sp=0x40005fb1e0 pc=0x6fa58c
github.com/syndtr/goleveldb/leveldb/table.(*Writer).Append(0x40000c0240, 0x40007e5441, 0x28, 0x31abbf, 0x40007e5469, 0x73, 0x31ab97, 0x100, 0x1)
        /home/kde/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/table/writer.go:255 +0x1e4 fp=0x40005fb2a0 sp=0x40005fb230 pc=0x6fa7f4
github.com/syndtr/goleveldb/leveldb.(*tWriter).append(0x40006c4060, 0x40007e5441, 0x28, 0x31abbf, 0x40007e5469, 0x73, 0x31ab97, 0x0, 0x0)
        /home/kde/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/table.go:559 +0xbc fp=0x40005fb310 sp=0x40005fb2a0 pc=0x71cb1c
github.com/syndtr/goleveldb/leveldb.(*tOps).createFrom(0x40002747e0, 0x15b0c88, 0x400016e000, 0x0, 0x0, 0x0, 0x0)
        /home/kde/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/table.go:397 +0x11c fp=0x40005fb3b0 sp=0x40005fb310 pc=0x71be9c
github.com/syndtr/goleveldb/leveldb.(*session).flushMemdb(0x40003782d0, 0x4000502000, 0x400002c620, 0x0, 0x0, 0x0, 0x0)
        /home/kde/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/session_compaction.go:35 +0xa4 fp=0x40005fb510 sp=0x40005fb3b0 pc=0x714114
github.com/syndtr/goleveldb/leveldb.(*DB).recoverJournal(0x400028e000, 0x400028e000, 0x40002bb0e0)
        /home/kde/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/db.go:612 +0xa80 fp=0x40005fb730 sp=0x40005fb510 pc=0x6ffde0
github.com/syndtr/goleveldb/leveldb.openDB(0x40003782d0, 0x0, 0x0, 0x40003782d0)
        /home/kde/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/db.go:131 +0x488 fp=0x40005fb820 sp=0x40005fb730 pc=0x6fe668
github.com/syndtr/goleveldb/leveldb.Open(0x15b0380, 0x4000208070, 0x40002b39b8, 0x0, 0x0, 0x0)
        /home/kde/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/db.go:203 +0xd0 fp=0x40005fb890 sp=0x40005fb820 pc=0x6fe8c0
github.com/syndtr/goleveldb/leveldb.OpenFile(0x400067e0f0, 0x26, 0x40002b39b8, 0x400019ed00, 0x4, 0x4)
        /home/kde/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/db.go:225 +0x7c fp=0x40005fb910 sp=0x40005fb890 pc=0x6feabc
github.com/ethereum/go-ethereum/ethdb/leveldb.New(0x400067e0f0, 0x26, 0x10, 0x10, 0x0, 0x0, 0x1, 0x400067e0f0, 0x26)
        /home/kde/meowsbits/core-geth/ethdb/leveldb/leveldb.go:98 +0x260 fp=0x40005fbb00 sp=0x40005fb910 pc=0x728fc0
github.com/ethereum/go-ethereum/core/rawdb.NewLevelDBDatabase(0x400067e0f0, 0x26, 0x0, 0x0, 0x0, 0x0, 0xa, 0x14, 0x4000060bc8, 0x46b1a8)
        /home/kde/meowsbits/core-geth/core/rawdb/database.go:299 +0x50 fp=0x40005fbb60 sp=0x40005fbb00 pc=0x858da0
github.com/ethereum/go-ethereum/cmd/geth.testDAOForkBlockNewChain(0x4000189380, 0x0, 0x0, 0x0, 0x40005af240, 0x481101)
        /home/kde/meowsbits/core-geth/cmd/geth/dao_test.go:110 +0x26c fp=0x40005fbeb0 sp=0x40005fbb60 pc=0xfd318c
github.com/ethereum/go-ethereum/cmd/geth.TestDAOForkBlockNewChain(0x4000189380)
        /home/kde/meowsbits/core-geth/cmd/geth/dao_test.go:87 +0x118 fp=0x40005fbf70 sp=0x40005fbeb0 pc=0xfd2e98
testing.tRunner(0x4000189380, 0x142f170)
        /usr/local/go/src/testing/testing.go:1194 +0xdc fp=0x40005fbfc0 sp=0x40005fbf70 pc=0x52811c
runtime.goexit()
        /usr/local/go/src/runtime/asm_arm64.s:1130 +0x4 fp=0x40005fbfc0 sp=0x40005fbfc0 pc=0x483d94
created by testing.(*T).Run
        /usr/local/go/src/testing/testing.go:1239 +0x244
goroutine 1 [chan receive]:            
testing.(*T).Run(0x4000189380, 0x13460c4, 0x18, 0x142f170, 0x6054ac01)    
        /usr/local/go/src/testing/testing.go:1240 +0x260        
testing.runTests.func1(0x4000001680)                             
        /usr/local/go/src/testing/testing.go:1512 +0x74         
testing.tRunner(0x4000001680, 0x40005fbd18)
        /usr/local/go/src/testing/testing.go:1194 +0xdc
testing.runTests(0x4000137fc8, 0x1d20cc0, 0x17, 0x17, 0xc00d497d996db4dd, 0x8bc182c769, 0x1f92920, 0x132cf00)
        /usr/local/go/src/testing/testing.go:1510 +0x29c        
testing.(*M).Run(0x400016fb80, 0x0)                              
        /usr/local/go/src/testing/testing.go:1418 +0x1b4        
github.com/ethereum/go-ethereum/cmd/geth.TestMain(0x400016fb80)
        /home/kde/meowsbits/core-geth/cmd/geth/run_test.go:64 +0x9c
main.main()                                                               
        _testmain.go:89 +0x160                                  
                                                                 
goroutine 37 [select]:                                          
github.com/ethereum/go-ethereum/consensus/ethash.(*remoteSealer).loop(0x40003781e0)
        /home/kde/meowsbits/core-geth/consensus/ethash/sealer.go:324 +0x154
created by github.com/ethereum/go-ethereum/consensus/ethash.startRemoteSealer
        /home/kde/meowsbits/core-geth/consensus/ethash/sealer.go:308 +0x21c                                                     
                                                         
goroutine 36 [chan receive]:                                                                                              
github.com/ethereum/go-ethereum/metrics.(*meterArbiter).tick(0x1f8aa20)
        /home/kde/meowsbits/core-geth/metrics/meter.go:290 +0x60
created by github.com/ethereum/go-ethereum/metrics.NewMeterForced         
        /home/kde/meowsbits/core-geth/metrics/meter.go:71 +0x10c                                                                   
                                                                 
goroutine 21 [chan receive, 1 minutes]:                                                                                             
github.com/ethereum/go-ethereum/core.(*txSenderCacher).cache(0x4000232bc0)
        /home/kde/meowsbits/core-geth/core/tx_cacher.go:63 +0x8c       
created by github.com/ethereum/go-ethereum/core.newTxSenderCacher      
        /home/kde/meowsbits/core-geth/core/tx_cacher.go:55 +0x9c
                                                        
goroutine 22 [chan receive, 1 minutes]:                         
github.com/ethereum/go-ethereum/core.(*txSenderCacher).cache(0x4000232bc0)
        /home/kde/meowsbits/core-geth/core/tx_cacher.go:63 +0x8c
created by github.com/ethereum/go-ethereum/core.newTxSenderCacher
        /home/kde/meowsbits/core-geth/core/tx_cacher.go:55 +0x9c
                                                                
goroutine 23 [chan receive, 1 minutes]:                         
github.com/ethereum/go-ethereum/core.(*txSenderCacher).cache(0x4000232bc0)
        /home/kde/meowsbits/core-geth/core/tx_cacher.go:63 +0x8c      
created by github.com/ethereum/go-ethereum/core.newTxSenderCacher      
        /home/kde/meowsbits/core-geth/core/tx_cacher.go:55 +0x9c       
                                                      
goroutine 24 [chan receive, 1 minutes]:                                
github.com/ethereum/go-ethereum/core.(*txSenderCacher).cache(0x4000232bc0)    
        /home/kde/meowsbits/core-geth/core/tx_cacher.go:63 +0x8c                               
created by github.com/ethereum/go-ethereum/core.newTxSenderCacher
        /home/kde/meowsbits/core-geth/core/tx_cacher.go:55 +0x9c

goroutine 274 [select]:
github.com/syndtr/goleveldb/leveldb.(*session).refLoop(0x40003782d0)
        /home/kde/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/session_util.go:189 +0x3b0
created by github.com/syndtr/goleveldb/leveldb.newSession
        /home/kde/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/session.go:93 +0x22c

goroutine 209 [select]:
github.com/syndtr/goleveldb/leveldb/util.(*BufferPool).drain(0x400002c540)
        /home/kde/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/util/buffer_pool.go:209 +0xcc
created by github.com/syndtr/goleveldb/leveldb/util.NewBufferPool
        /home/kde/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/util/buffer_pool.go:240 +0x154
FAIL    github.com/ethereum/go-ethereum/cmd/geth        81.912s
?       github.com/ethereum/go-ethereum/cmd/p2psim      [no test files]
?       github.com/ethereum/go-ethereum/cmd/rlpdump     [no test files]
ok      github.com/ethereum/go-ethereum/cmd/utils       0.260s
```

### PR DESCRIPTION
**Pull request by meowsbits**
This fixes #346, and includes some extras.

- The fix: 69c9cc0
- Other are extras, but related and good to have overall I think. (Except dd4fea0, the rest are commits that we'll merge in eventually anyway).

The cherry-picked `go.mod` changes required conflict resolution.

### CODE REVIEW

File: cmd/geth/accountcmd_test.go [hunk 1]
```diff
@@ -43,13 +43,13 @@ func tmpDatadirWithKeystore(t *testing.T) string {
 }
 
 func TestAccountListEmpty(t *testing.T) {
-	geth := runGeth(t, "account", "list")
+	geth := runGeth(t, "--nousb", "account", "list")
```
Comment (ziogaschr) [inline cmd/geth/accountcmd_test.go]: I might remove the `--nousb` if not really needed, as in the next go-ethereum merge, the revert the flag to `--usb`

File: cmd/geth/consolecmd.go [hunk 2]
```diff
@@ -231,13 +229,10 @@ func ephemeralConsole(ctx *cli.Context) error {
 			utils.Fatalf("Failed to execute %s: %v", file, err)
 		}
 	}
-	// Wait for pending callbacks, but stop for Ctrl-C.
-	abort := make(chan os.Signal, 1)
-	signal.Notify(abort, syscall.SIGINT, syscall.SIGTERM)
 
 	go func() {
-		<-abort
-		os.Exit(0)
```
Comment (ziogaschr) [inline cmd/geth/consolecmd.go:234]: Was the issue indeed the `os.Exit(0)` on latest go?

Comment (meowsbits) [reply cmd/geth/consolecmd.go:234]: No... it was the `snappy` library. https://github.com/etclabscore/core-geth/commit/69c9cc008d4b742dddb31f138b201547e91812ab

File: cmd/geth/accountcmd_test.go (hunk 1 above)
Comment (meowsbits) [reply cmd/geth/accountcmd_test.go]: Done with https://github.com/etclabscore/core-geth/pull/347/commits/c5fe6a635aab6df080197113dd078dfd7eec9603

**Review by ziogaschr — APPROVED** [review]
LGTM! I don't have an ARM64 device to give it a test, so I trust you and @iquidus for testing it.

**Comment by meowsbits** [comment]
:+1:

> for testing it.

Also Travis :wink:

**Comment by ziogaschr** [comment]
Checked Travis, but in general I give a local test on reviewing stuff as well

**Pull request merged**